src/chocolatey/infrastructure.app/services/ChocolateyPackageService.cs

```diff
@@ -550,6 +552,11 @@ public void upgrade_noop(ChocolateyConfiguration config)
             return packageUpgrades;
         }
 
+        private void before_package_upgrade(PackageResult packageResult, ChocolateyConfiguration config)
+        {
+            _powershellService.run_package_scripts_for_command(config, packageResult, CommandNameType.upgrade);
+        }
+

```

Comment:
Integration tests now verify the following in UpgradeScenarios:
1) The modifications script from the initial package is run.
2) The modifications script from the new package is not run.
3) The installation script from the new package is run.

They also verify that in UninstallScenarios:
1) The modifications script from the installed package is run.
2) The uninstall script from the installed package is run.

I believe that the current functionality does match what is desired.